drivers/misc/cldatadrv/rcar_cluster_drv.c

```diff
@@ -0,0 +1,389 @@
+{
+	rcar_cluster_device_t *clusterdvc = NULL;
+	int ret = 0;
+	struct miscdevice *dev = NULL;
+
+	dev_info(&rpdev->dev, "cluster: send_msg: probe\n");
+
+	clusterdvc = devm_kzalloc(&rpdev->dev, sizeof(*clusterdvc), GFP_KERNEL);
+
+	if (clusterdvc == NULL)
+		return -ENOMEM;
+
+	clusterdvc->rpdev = rpdev;
+	dev = &clusterdvc->dev;
+	dev->parent = &rpdev->dev;
+	dev->minor = MISC_DYNAMIC_MINOR;
+	dev->name = kstrdup("cluster-taurus", GFP_KERNEL);
+
+	if (!dev->name) {
+		ret = -ENOMEM;
+		goto free_name;
+	}
+
+	dev->fops = &misc_fops;
+	ret = misc_register(dev);
+
+	dev_set_drvdata(&rpdev->dev, clusterdvc);
+
+	INIT_LIST_HEAD(&clusterdvc->taurus_event_list_head);
+	rwlock_init(&clusterdvc->event_list_lock);
+
+	return ret;
+
+free_name:
+	kfree(dev->name);
+	dev->name = NULL;
+
+free_cluster_data:
+	devm_kfree(&rpdev->dev, clusterdvc);
+
+	return ret;
+}
+
+static void rpmsg_cluster_remove(struct rpmsg_device *rpdev)
+{
+	rcar_cluster_device_t *data = dev_get_drvdata(&rpdev->dev);
+
+	misc_deregister(&data->dev);
+
+	kfree(data->dev.name);
+	data->dev.name = NULL;
+
+	devm_kfree(&rpdev->dev, data);
+}
+
+
+static int __init cluster_drv_init(void)
+{
+	int ret = 0;
+
+	ret = register_rpmsg_driver(&rpmsg_cluster_drv);
+	if (ret < 0) {
+		pr_err("failed to register %s driver\n", __func__);
+		return -EAGAIN;
+	}
+
+	return ret;
+}
+late_initcall(cluster_drv_init);
+
+static void __exit cluster_drv_exit(void)
+{
+	unregister_rpmsg_driver(&rpmsg_cluster_drv);
+}
+
+static int misc_open(struct inode *inode, struct file *p_file)
+{
+	struct miscdevice *misc_device = p_file->private_data;
+	rcar_cluster_device_t *cldev = container_of(misc_device, struct rcar_cluster_device, dev);
+
+	p_file->private_data = cldev;
+
+	return 0;
+}
+
+static int misc_release(struct inode *inode, struct file *p_file)
+{
+	return 0;
+}
+
+static long misc_ioctl(struct file *p_file, unsigned int cmd, unsigned long arg)
+{
+	unsigned int available_commands[] = {
+		CLUSTER_SPEED,
+		CLUSTER_GEAR,
+		CLUSTER_RPM,
+		CLUSTER_TURN,
+		CLUSTER_DOOR_OPEN,
+		CLUSTER_FOG_LIGHTS_BACK,
+		CLUSTER_FOG_LIGHTS_FRONT,
+		CLUSTER_HIGH_BEAMS_LIGHT,
+		CLUSTER_HIGH_ENGINE_TEMPERATURE,
+		CLUSTER_LOW_BATTERY,
+		CLUSTER_LOW_BEAMS_LIGHTS,
+		CLUSTER_LOW_FUEL,
+		CLUSTER_LOW_OIL,
+		CLUSTER_LOW_TIRE_PRESSURE,
+		CLUSTER_SEAT_BELT,
+		CLUSTER_SIDE_LIGHTS,
+		CLUSTER_BATTERY_ISSUE,
+		CLUSTER_AUTO_LIGHTING_ON,
+		CLUSTER_ACTIVE
+	};
+	unsigned int index = 0;
+	unsigned int len = sizeof(available_commands)/sizeof(unsigned int);
+	rcar_cluster_device_t *cldev = p_file->private_data;
+	taurus_cluster_data_t *data = NULL;
+	taurus_cluster_res_msg_t res;
+	void *kbuf = NULL;
+
+	/*verify of command is supported*/
+	for (index = 0; index < len ; ++index) {
+		if (available_commands[index] == cmd)
+			break;
+	}
+
+	if (index == len)
+		return -EINVAL;
+
+	kbuf = kzalloc(sizeof(taurus_cluster_data_t), GFP_KERNEL);
+	if (!kbuf)
+		return -ENOMEM;
+
+	data = (taurus_cluster_data_t *)kbuf;
```

Comment:
Why do you need `kbuf`? Why not 

`data = kzalloc(sizeof(taurus_cluster_data_t), GFP_KERNEL);`?

Reply:
Also, this is short-lived small buffer. You can allocate in right on stack.

Reply:
Moreover, you can remove `sizeof(taurus_cluster_data_t)` at all. Just pass `cmd` and `arg` as a two separate parameters to `send_msg`.  You will unpack this structure there anyways. No need for extra memory operations.